### PR TITLE
SDHC driver

### LINKED ISSUE #6: SDIO driver
The SD bus is based on an old Toshiba design and is basically the same as the one in the DSi, except with a ~67MHz clock rather than ~33MHz. It's well documented in [gbatek](https://problemkaputt.de/gbatek.htm#dsisdmmcprotocolandioports) and is a piece necessary to get WiFi working.

It could be based off the preexisting toshsd or tmio drivers with a few modifications, since our version of the chip has the separate 32bit FIFO.

[x] base driver
[] 32bit transfers
[x?] full SDIO support

### PR DESCRIPTION
**Pull request by Wolfvak**
this is very much a WIP to add wifi support: the chip is connected via SDIO to the 3DS and it's identified as an "AR6014G". this PR will add a working SD bus driver and wifi

fixes #6

### CODE REVIEW

**Comment by nickdesaulniers** [comment]
From userspace, how do we usually configure wifi?  Is it possible with what's in toybox via buildroot?

**Comment by Wolfvak** [comment]
right now the ath6kl driver isn't really working, nor do I really know if it's supposed to without modifications
in the future we'll need to add scripts to automate the task, since it can get quite hairy between iwconfig, ip and wpa_supplicant

**Comment by ParzivalWolfram** [comment]
Is this the same flaky SDHC support seen in older builds or is it stable
enough to support writes?

**Comment by Wolfvak** [comment]
This is a proper SD bus driver, it accesses the second controller which is connected only to the WiFi card. This was tested by temporarily redirecting the SD card port to controller 3 and adding support for that. This is capable of exposing the SDIO card, its class, device IDs and handling the SDIO interrupt pin.

**Comment by Wolfvak** [comment]
ok after adding support for the "secure cmd" stuff it gets farther through the ath6kl driver but still dies with
`ath6kl: Unsupported hardware version: 0x2300006f`

this was on an old3DS XL so it's likely other consoles will get different IDs, the entries will need to be added to https://github.com/linux-3ds/linux/blob/048def1d291e0533426d654ce79a918b2689516f/drivers/net/wireless/ath/ath6kl/init.c#L35

**Comment by ParzivalWolfram** [comment]
I can add mine from my N3DSXL once I get it compiled.

**Comment by Wolfvak** [comment]
Currently I dont even know how to add them properly (there's multiple data fields), plus it seems the ath6kl driver doesn't support the AR6002 which is the module in all DSi and 3DS models.

So WiFi might have to wait for a bit.

**Comment by ParzivalWolfram** [comment]
Still might be useful to see if they differ, or have different model
numbers at least.

**Comment by Wolfvak** [comment]
ok I'm changing this PR to SD Host Controller driver only and leave wifi for sometime later, it needs a bit more research into how the ath6kl driver works

File: drivers/platform/nintendo3ds/ctr_sdhc.h [hunk 1]
```diff
@@ -0,0 +1,147 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ *  Nintendo 3DS Secure Digital Host Controller driver
+ *
+ *  Copyright (C) 2021 Santiago Herrera
+ */
+
+#define SDHC_HCLK	67027964	/* ~67MHz */
```
Comment (xerpi) [inline drivers/platform/nintendo3ds/ctr_sdhc.h:8]: Should this come from the device tree?
And does it come from a configurable PLL or some clockgen device?

Comment (Wolfvak) [reply drivers/platform/nintendo3ds/ctr_sdhc.h:8]: probably yeah, it's the same freq as the NDS clk (or 1/2 the DSi cpu clk, or 1/4 the 3DS cpu clk)
other devices might also use it, so I'll just add more clocks derivated from the 133MHz clock (*2 and /2)

**Comment by xerpi** [comment]
Should `ctr_sdhc.c` be moved to `drivers/mmc/host`?
And I think `ctr_sdhc.h` is not needed since it's not exposing any API, we can just move all the reg defines to the .c. I think that's a common way of doing things in Linux, check for example https://elixir.bootlin.com/linux/latest/source/drivers/mmc/host/sdhci-tegra.c

**Comment by Wolfvak** [comment]
I'm not sure, this driver is definitely 3DS specific (and DSi for that matter) so it's not too out of place in the platform-specific drivers section.

As for the header thing, there's so many register definitions I thought it'd be cleaner to separate them from the driver itself. This is also what the original toshsd driver did.

File: drivers/net/wireless/ath/ath6kl/sdio.c [hunk 2]
```diff
@@ -1408,6 +1408,7 @@ static void ath6kl_sdio_remove(struct sdio_func *func)
 }
 
 static const struct sdio_device_id ath6kl_sdio_devices[] = {
+	{SDIO_DEVICE(SDIO_VENDOR_ID_ATHEROS, 0x201)}, /* DSi/3DS wifi chip ID? */
```
Comment (nickdesaulniers) [inline drivers/net/wireless/ath/ath6kl/sdio.c:1411]: Do we want to leave this out until we have the base SDHC work landed?

File: drivers/platform/nintendo3ds/Kconfig [hunk 3]
```diff
@@ -15,6 +15,13 @@ menuconfig NINTENDO3DS_PLATFORM_DEVICES
 
 if NINTENDO3DS_PLATFORM_DEVICES
 
+config CTR_SDHC
+	tristate "Nintendo 3DS SD Host Controller"
+	select MMC
+	default y
+	help
+	  TODO
```
Comment (nickdesaulniers) [inline drivers/platform/nintendo3ds/Kconfig:23]: Worth filling in.

File: drivers/net/wireless/ath/ath6kl/sdio.c (hunk 2 above)
Comment (Wolfvak) [reply drivers/net/wireless/ath/ath6kl/sdio.c:1411]: Nope, that was just an added ID to get stuff tested, will remove

File: drivers/platform/nintendo3ds/Kconfig (hunk 3 above)
Comment (Wolfvak) [reply drivers/platform/nintendo3ds/Kconfig:23]: Yeah, I'm just terrible at descriptions but will think of something to put in there

**Comment by Wolfvak** [comment]
I also have to mention this driver doesnt perform to its best, as it doesnt use DMA, or even the 32bit FIFO at all but it's more important to get something that works correctly first, after wifi support is added to the ath6kl driver we can start worrying about improving performance using actual benchmark data.

**Review by nickdesaulniers — APPROVED** [review]

**Comment by Wolfvak** [comment]
Oh and another side note related to the wifi chip specifically: it needs to be powered on through GPIO4.pin0

**Pull request merged**